### ISSUE TITLE
undefined payload template variable

### ISSUE THREAD

**Issue by lauthrul** [emqx/emqx#3762]
**Environment**:

- EMQ X version: 4.2.1
- OS: CentOS 7
- Kernel: Linux dj-dev 3.10.0-1062.18.1.el7.x86_64 #1 SMP Tue Mar 17 23:49:17 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

**What happened and what you expected to happen**:
Get `undefined` payload template variable for Rule action. Expect the exact value of these variables.
- Here is Rule SQL:
```
SELECT
  payload
FROM
  "sensor/data"
```

- Action Handler:
```
Type: republish
target_topic: repub/to/${clientid}
target_qos: 0
payload_tmpl: ${payload.temperature}, ${payload.humidity}, ${payload.volume}, ${payload.PM10}, ${payload.pm25}, ${payload.SO2}, ${payload.NO2}, ${payload.CO}, '${payload.id}', ${payload.area}, ${payload.ts}

```

- MQTT payload:
```
{
    "temperature": 30,
    "humidity" : 20,
    "volume": 44.5,
    "PM10": 23,
    "pm25": 61,
    "SO2": 14,
    "NO2": 4,
    "CO": 5,
    "id": "10-c6-1f-1a-1f-47",
    "area": 1,
    "ts": 1596157444170
}
```

- What I get:
```
Topic: repub/to/undefined QoS: 0

undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined, 'undefined', undefined, undefined
```

**Comment by terry-xiaoyu** [comment]
Try to use function json_decode() in the rule sql:

```
SELECT
  json_encode(payload) as payload
FROM
  "sensor/data"
```

https://docs.emqx.net/broker/latest/cn/rule/rule-engine.html#编解码函数

Note that `payload.k` is only a syntactic sugar of `map_get('k', json_encode(payload))` and can only be used in the rule sql.

**Comment by lauthrul** [comment]
@terry-xiaoyu thanks for reply. this works:
```
SELECT
  json_decode(payload) as payload
FROM
  "sensor/data"
```
but ${clientid} still undefined. (see the topic from action handler result below)
```
Topic: repub/to/undefined QoS: 0

30, 20, 44.5, 23, 61, 14, 4, 5, '10-c6-1f-1a-1f-47', 1, 1596157444170
```

by the way,  `payload.k` is available even no json_decode(payload) on version 4.1.2.

**Comment by terry-xiaoyu** [comment]
For the ${clientid}, you should specify it explicitly in the SELECT clause:

```
SELECT
  json_encode(payload) as payload,
  clientid
FROM
  ...
```

or just use `*`:

```
SELECT
  json_encode(payload) as payload,
  *
FROM
  ...
```

If `payload.k` was supported in earlier versions then it's a bug and I will fix it in emqx 4.2.2.

**Comment by lauthrul** [comment]
@terry-xiaoyu ok, thanks a lot!
